src/libexpr/eval.cc

```diff
@@ -821,8 +821,17 @@ inline Value * EvalState::lookupVar(Env * env, const ExprVar & var, bool noEval)
 
 Value * EvalState::allocValue()
 {
+    if (!valueAllocCache) {
+        valueAllocCache = GC_malloc_many(sizeof(Value));
+        if (!valueAllocCache) throw std::bad_alloc();
+    }
+
+    void * p = valueAllocCache;
+    GC_PTR_STORE_AND_DIRTY(&valueAllocCache, GC_NEXT(p));
+    GC_NEXT(p) = nullptr;
```

Comment:
Isn’t this line equivalent to `valueAllocCache = nullptr`? I guess it’s not because that would make the whole thing quite inefficient, but I don’t see how it isn’t

Reply:
nope, because the line before is equivalent to `valueAllocCache = *valueAllocCache`. boehm malloc_many returns a list of objects in the form of a "first word is next pointer". it also explicitly says to clear the next pointer of an object taken from the list, which is what this line does.

Reply:
Oh yes sorry, I misread the thing. Makes sense now, thanks